lib/client/rpc/openbookMarkets.ts

```diff
@@ -0,0 +1,401 @@
+      order.side === "ask"
+        ? market.marketInstance.account.baseMint
+        : market.marketInstance.account.quoteMint;
+    const marketVault =
+      order.side === "ask"
+        ? market.marketInstance.account.marketBaseVault
+        : market.marketInstance.account.marketQuoteVault;
+    // TODO not sure how this works when clientOrderId not specified
+    const clientOrderId = new BN(order.clientOrderId || Date.now());
+    const placeOrderArgs = this.createPlaceOrderArgs({
+      accountIndex: clientOrderId,
+      market: market.marketInstance,
+      isAsk: order.side === "ask",
+      amount: order.size,
+      price: order.price,
+      isLimitOrder: placeOrderType === "limit",
+      isPostOnlyOrder: false,
+    });
+    if (placeOrderArgs) {
+      const placeTx = await this.openbookTwap.methods
+        .placeOrder(placeOrderArgs)
+        .accounts({
+          openOrdersAccount: openOrdersAccountAddress,
+          asks: market.marketInstance.account.asks,
+          bids: market.marketInstance.account.bids,
+          eventHeap: market.marketInstance.account.eventHeap,
+          market: market.publicKey,
+          marketVault,
+          twapMarket: getTwapMarketKey(
+            market.publicKey,
+            this.openbookTwap.programId
+          ),
+          userTokenAccount: getAssociatedTokenAddressSync(
+            mint,
+            this.transactionSender.owner,
+            true
+          ),
+          openbookProgram: this.openbook.programId,
+        })
+        .transaction();
+      return this.transactionSender.send(
+        [placeTx],
+        this.rpcProvider.connection
+      );
+    }
+    return [];
+  }
+
+  private createOrderToPlace(
+    order: Omit<OpenbookOrder, "status">,
+    placeOrderType: PlaceOrderType
+  ): OrderToPlace {
+    // TODO support more types
+    const orderType =
+      placeOrderType === "limit"
+        ? PlaceOrderTypeUtils.Limit
+        : PlaceOrderTypeUtils.Market;
+    const orderToPlace: OrderToPlace = {
+      price: order.price,
+      side: order.side === "ask" ? SideUtils.Ask : SideUtils.Bid,
+      size: order.size,
+      clientOrderId: order.clientOrderId,
+      expiryTimestamp: new BN(0),
+      selfTradeBehavior: SelfTradeBehaviorUtils.AbortTransaction,
+      orderType,
+    };
+    return orderToPlace;
+  }
+
+  private createPlaceOrderArgs({
+    amount,
+    price,
+    isLimitOrder,
+    isPostOnlyOrder,
+    isAsk,
+    accountIndex,
+    market,
+  }: {
+    amount: number;
+    price: number;
+    isLimitOrder?: boolean;
+    isPostOnlyOrder?: boolean;
+    isAsk?: boolean;
+    accountIndex: number;
+    market: OBMarket;
+  }): TwapPlaceOrderArgs | undefined {
+    let priceLots = market.priceLotsToUi(price);
+    const _priceLots = market.priceLotsToUi(price);
+    const maxBaseLots = market.baseLotsToUi(amount);
+    let maxQuoteLotsIncludingFees = market.quoteLotsToUi(
+      priceLots * maxBaseLots
+    );
+
+    if (!isLimitOrder) {
```

Comment:
we want to use placeTakeOrder if we're not dealing with a limit.

Reply:
this is handled